### DISCUSSION TITLE
Attaching a refreshed certificate

### DISCUSSION THREAD

**Discussion by denchen** [chimurai/http-proxy-middleware#575] [Q&A]
### Is this a question?

I asked this [over on StackOverflow but haven't gotten a response](https://stackoverflow.com/questions/53161380/http-proxy-middleware-refreshing-certificate)

I'm using this library to create a proxy to a server that requires certificates (`cert`, `ca`, and `key`). The problem is, these certs expire over time, so I need to retrieve new certificates. However, I'm unsure how to attach a new certificate to an existing proxy. 

My basic setup is similar to the following:

```
app.use(
  '/api',
  proxy({
    target: {
      protocol: 'https:',
      host: 'another.domain.com',
      port: 443,
      cert: cert,
      ca: ca_cert,
      key: key
    },
    secure: false,
    changeOrigin: true
  })
);
```

Is it possible to insert a middleware before the `proxy()` to retrieve new certs and have `proxy()` use the new certs? Something like

```
app.use(
  '/api',
  (req, res, next) => {
    if (/* cert has expired */) {
      cert = /* get new cert */
    }
  },
  proxy({
    target: {
      protocol: 'https:',
      host: 'another.domain.com',
      port: 443,
      cert: cert,
      ca: ca_cert,
      key: key
    },
    secure: false,
    changeOrigin: true
  })
);
```

Or will that not work? Can I use `onProxyReq()` to manipulate the certs? Or can I tear down the original `proxy()` instance and instantiate a new one?

### Setup

* http-proxy-middleware: _0.19.0_
* http-proxy-middleware configuration
* server: _express@4.16.4_

**Comment by iverberk** [comment]
I know this is old but we have a similar use-case. We need to fetch/check the client certificates on-the-fly (when the request is made) and they are only valid for five minutes. Statically configuring the certificates won't work for us. Is there any way that we can alter/set the client certificates in the onProxyReq event handler?

@chimurai can you please tell us if this is (im)possible? If it is not possible we need to rethink our solution.

**Reply by chimurai** [reply]
You could try the `router` option. ([router recipes](https://github.com/chimurai/http-proxy-middleware/blob/master/recipes/router.md#custom-router-function))

It'll allow you to return a dynamic `target`; Before returning the `target` you can fetch/check certs and configure them in the target object.

```javascript
// Asynchronous router function which returns promise
router: async function(req) {
    const target = await doSomeIO();
    return target;
}
```

`target` can be an string or an [target object](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/92af6adfa92bac7261e17aefeb93a7fb2d44317b/types/http-proxy/index.d.ts#L20-L33).

**Reply by iverberk** [reply]
Thanks a lot @chimurai, your suggestion works like a charm!
